api/api-stage-operations-v1/src/main/java/com/thoughtworks/go/apiv1/stageoperations/StageOperationsControllerV1Delegate.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright 2018 ThoughtWorks, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.thoughtworks.go.apiv1.stageoperations;
+
+import com.thoughtworks.go.api.ApiController;
+import com.thoughtworks.go.api.ApiVersion;
+import com.thoughtworks.go.api.spring.ApiAuthenticationHelper;
+import com.thoughtworks.go.api.util.HaltApiResponses;
+import com.thoughtworks.go.api.util.MessageJson;
+import com.thoughtworks.go.server.service.ScheduleService;
+import com.thoughtworks.go.spark.Routes;
+import org.apache.commons.lang.StringUtils;
+import org.springframework.http.HttpStatus;
+import spark.Request;
+import spark.Response;
+
+import java.util.Arrays;
+
+import static java.lang.String.format;
+import static spark.Spark.*;
+
+public class StageOperationsControllerV1Delegate extends ApiController {
+    private static final int ACCEPTED = HttpStatus.ACCEPTED.value();
+
+    private final ScheduleService scheduleService;
+    private final ApiAuthenticationHelper apiAuthenticationHelper;
+
+    public StageOperationsControllerV1Delegate(ScheduleService scheduleService, ApiAuthenticationHelper apiAuthenticationHelper) {
+        super(ApiVersion.v1);
+        this.scheduleService = scheduleService;
+        this.apiAuthenticationHelper = apiAuthenticationHelper;
+    }
+
+    @Override
+    public String controllerBasePath() {
+        return Routes.Stage.BASE;
+    }
+
+    @Override
+    public void setupRoutes() {
+        path(controllerPath(), () -> {
+            before("", mimeType, this::setContentType);
+            before("/*", mimeType, this::setContentType);
+            before("", this::verifyContentType);
+            before("/*", this::verifyContentType);
+
+            before(Routes.Stage.TRIGGER_PATH, mimeType, apiAuthenticationHelper::checkPipelineGroupOperateUserAnd403);
+
+            post(Routes.Stage.TRIGGER_PATH, mimeType, this::run);
+        });
+    }
+
+    private String run(Request req, Response res) {
+        String pipelineName = req.params("pipeline_name");
+        String pipelineCounter = req.params("pipeline_counter");
+        String stageName = req.params("stage_name");
+
+        try {
+            scheduleService.rerunStage(pipelineName, pipelineCounter, stageName);
+        } catch (Exception e) {
+            HaltApiResponses.haltBecauseOfReason(e.getMessage());
```

Comment:
Catching all exceptions and sending the message in the response seems wrong. But it looks like the service throws a bunch of RuntimeExceptions in scenarios that it checks for. Does anyone have any suggestions if this can be avoided? @ketan ?

Reply:
@akshaydewan likely you can do one of two things without making big changes: 

1. Test for the message formats. Yuck.
2. Make specific exceptions that extend RuntimeException and test for those.